generators/chipyard/src/main/scala/IOBinders.scala

```diff
@@ -65,74 +79,121 @@ class ComposeIOBinder[T](fn: => (Clock, Bool, Bool, T) => Seq[Any])(implicit tag
 
 // DOC include end: IOBinders
 
+object AddIOCells {
+  def gpio(gpios: Seq[GPIOPortIO]): (Seq[Seq[Analog]], Seq[Seq[IOCell]]) = {
+    gpios.zipWithIndex.map({ case (gpio, i) =>
+      gpio.pins.zipWithIndex.map({ case (pin, j) =>
+        val g = IO(Analog(1.W))
+        // TODO functionalize this?
+        val iocell = IOCell.exampleGPIO()
+        iocell.suggestName(s"iocell_gpio_${i}_${j}")
+        iocell.io.o := pin.o.oval
+        iocell.io.oe := pin.o.oe
+        iocell.io.ie := pin.o.ie
+        pin.i.ival := iocell.io.i
+        iocell.io.pad <> g
+        (g, iocell)
+      }).unzip
+    }).unzip
+  }
+
+  def uart(uartPins: Seq[UARTPortIO]): (Seq[UARTPortIO], Seq[Seq[IOCell]]) = {
+    uartPins.zipWithIndex.map({ case (u, i) =>
+      IOCell.generateIOFromSignal(u, Some(s"iocell_uart_${i}"))
+    }).unzip
+  }
+
+  def debug(psd: PSDIO, debugOpt: Option[DebugIO]): (PSDIO, Option[DebugIO], Seq[IOCell]) = {
+    val (psdPort, psdIOs) = IOCell.generateIOFromSignal(psd, Some("iocell_psd"))
+    val optTuple = debugOpt.map(d => IOCell.generateIOFromSignal(d, Some("iocell_debug")))
+    val debugPortOpt: Option[DebugIO] = optTuple.map(_._1)
+    val debugIOs: Seq[IOCell] = optTuple.map(_._2).toSeq.flatten
+    (psdPort, debugPortOpt, psdIOs ++ debugIOs)
+  }
+
+  def serial(serial: SerialIO): (SerialIO, Seq[IOCell]) = {
+    IOCell.generateIOFromSignal(serial, Some("iocell_serial"))
+  }
+}
+
 class WithGPIOTiedOff extends OverrideIOBinder({
-  (c, r, s, top: HasPeripheryGPIOModuleImp) => top.gpio.map(gpio => gpio.pins.map(p => p.i.ival := false.B)); Nil
+  (system: HasPeripheryGPIOModuleImp) => {
+    // TODO: Do we suggestName the ports?
+    val (ports2d, ioCells2d) = AddIOCells.gpio(system.gpio)
+    val harnessFn = (th: TestHarness) => { ports2d.flatten.foreach(_ <> AnalogConst(0)); Nil }
+    Seq((ports2d.flatten, ioCells2d.flatten, Some(harnessFn)))
+  }
 })
 
+class WithUARTAdapter extends OverrideIOBinder({
+  (system: HasPeripheryUARTModuleImp) => {
+    // TODO: Do we suggestName the ports?
+    val (ports, ioCells2d) = AddIOCells.uart(system.uart)
+    val harnessFn = (th: TestHarness) => { UARTAdapter.connect(ports)(system.p); Nil }
+    Seq((ports, ioCells2d.flatten, Some(harnessFn)))
+  }
+})
+
+// TODO: Add a note about synthesizability of SimBlockDevice
 class WithSimBlockDevice extends OverrideIOBinder({
-  (c, r, s, top: CanHavePeripheryBlockDeviceModuleImp) => top.connectSimBlockDevice(c, r); Nil
+  (system: CanHavePeripheryBlockDeviceModuleImp) => SimBlockDevice.connect(system.clock, system.reset.asBool, system.bdev)(system.p); Nil
 })
```

Comment:
Can we revert to `top.connectSimBlockDevice` here?

Reply:
Sure.

Reply:
This is done